feat: configs for hparam optimization (#34)
thousandbrainsproject/tbp.tbs_sensorimotor_intelligence: pytorch/configs/experiment/01_hyperparameter_optimization/README.md

## 📝 Starting Configuration (`starting_config.yaml`)

The starting configuration defines the following:

- RGBD_ViT model with `vit-b16-224-in21k` backbone from HuggingFace with the following modifications:
  - Two "simple" heads (`nn.Linear`) for classification and quaternion prediction
  - Full training of all parameters enabled
  - RGBD patch embedding initialized from pretrained RGB weights + mean of the pretrained RGB weights for Depth Channel:

```python
    new_projection = nn.Conv2d(in_channels=4,
        out_channels=out_channels,
        kernel_size=kernel_size,
        stride=stride,
        padding=padding,
    )
    # Copy RGB weights and initialize depth channel
    with torch.no_grad():
        depth_weight = rgb_weights.mean(dim=1, keepdim=True)  # (out_c, 1, k, k)
        new_weight = torch.cat(
            [rgb_weights, depth_weight], dim=1
        )
```

- Dataset and DataLoaders:

>   - Dataset contains 1,078 images (77 objects x 14 rotations), randomly split into train/val with 80/20 ratio

Reviewer: ```suggestion
  - Dataset contains 1,078 images (77 objects x 14 rotations), randomly split into train/val with 80/20 ratio during hyperparameter optimization
```

  - Without Early Stopping: Training becomes unstable and model diverges
- **Decision**: Keep Early Stopping enabled
- **Impact**: Reduces training FLOPs by ~3x (73 epochs vs 200) but maintains model quality

### 2. Gradient Clipping

- **Reason**: Regularize ViT model and prevent divergence
- **Results**:
  - With gradient clipping (1.0): `val/class_acc` = 77.3%; `test/class_acc` = 76.1%
- **Decision**: Enable gradient clipping with value 1.0

### 3. Freezing Backbone

- **Reason**: Common practice in finetuning to freeze backbone or early layers
- **Results**:
  - With frozen backbone: `val/class_acc` = 76.9%; `test/class_acc` = 71.6%
- **Decision**: Keep all parameters trainable

### 4. Data Augmentation

- **Reason**: Standard practice for finetuning on small datasets
- **Implementation**: Only rotation-invariant augmentations:

```python
self.rgb_transform: T.Compose = T.Compose(
    [
        T.RandomResizedCrop(224, scale=(0.9, 1.0), ratio=(0.9, 1.1)),
        T.ColorJitter(0.2, 0.2, 0.2, 0.05),
        T.Normalize(mean=[0.5, 0.5, 0.5], std=[0.5, 0.5, 0.5]),
    ]
)
```

- **Results**:
  - With augmentation: `val/class_acc` = 87.5%; `test/class_acc` = 78.4%

> - **Final Decision**: Omit data augmentation to match training conditions with other models

Reviewer: ```suggestion
- **Final Decision**: Omit data augmentation to match training conditions with Monty models (receive no data-augmentation, and are restricted from using their optimal exploration policies that would acquire more data)
```

### 6. Warmup + Cosine Decay LR Schedule

- **Reason**: More standard in ViT finetuning than ReduceLROnPlateau
- **Results**:
  - `val/class_acc` = 88.4%; `test/class_acc` = 80.5%
- **Decision**: Keep Warmup + Cosine Decay schedule

### 7. AdamW Optimizer

- **Reason**: Standard practice, properly applies weight decay unlike Adam
- **Results**:
  - `val/class_acc` = 90.3%; `test/class_acc` = 86.8%
- **Decision**: Use AdamW

### 8. Differential Learning Rates

- **Reason**: Common to use lower LR for pretrained backbone vs new heads
- **Results**:
  - `val/class_acc` = 87.5%; `test/class_acc` = 86.0%
- **Decision**: Use same learning rate for all parameters

### 9. Loss Weight Tuning ($\\lambda\_\\text{rot}$)

- **Reason**: Validate empirical choice of $\\lambda\_\\text{rot} = 1.0$
- **Results**:
  - $\\lambda\_\\text{rot} = 0.1$: `val/class_acc = 92.6%`, `val/rotation_error = 113.7°`
  - $\\lambda\_\\text{rot} = 1.0$: `val/class_acc = 90.3%`, `val/rotation_error = 109.2°`
  - $\\lambda\_\\text{rot} = 3.0$: `val/class_acc = 89.4%`, `val/rotation_error = 106.2°`
  - $\\lambda\_\\text{rot} = 5.0$: `val/class_acc = 88.9%`, `val/rotation_error = 105.8°`

> - **Decision**: Keep $\\lambda\_\\text{rot} = 1.0$

Reviewer: ```suggestion
- **Decision**: Keep $\\lambda\_\\text{rot} = 1.0$ to balance accuracy and rotation error performance.
```
